File: infra/lambda/agent-api/handler.py
"""
Agent API Lambda Handler

Multimodal / Voice エージェント。
Bedrock Nova API を使用して実際の画像解析・音声処理を実行。

CORS 対応済み。
"""

import json
import logging
import os
import base64
import boto3
from botocore.config import Config

logger = logging.getLogger()
logger.setLevel(os.environ.get("LOG_LEVEL", "INFO"))

# AWS リージョン
BEDROCK_REGION = os.environ.get("BEDROCK_REGION", "us-east-1")

# Bedrock クライアント初期化
bedrock_config = Config(
    region_name=BEDROCK_REGION,
    retries={"max_attempts": 3, "mode": "adaptive"},
)
bedrock_runtime = boto3.client("bedrock-runtime", config=bedrock_config)

# モデル ID
NOVA_LITE_MODEL = os.environ.get("NOVA_LITE_MODEL", "amazon.nova-lite-v1:0")
NOVA_PRO_MODEL = os.environ.get("NOVA_PRO_MODEL", "amazon.nova-pro-v1:0")

# CORS ヘッダー
CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type,Authorization,X-Amz-Date",
    "Access-Control-Allow-Methods": "GET,POST,PUT,DELETE,OPTIONS",
}


def lambda_handler(event, context):
    """Lambda ハンドラー"""
    logger.info(f"Event: {json.dumps(event)[:500]}")

    http_method = event.get("httpMethod", "GET")
    path = event.get("path", "")

    # OPTIONS (CORS preflight)
    if http_method == "OPTIONS":
        return {
            "statusCode": 200,
            "headers": CORS_HEADERS,
            "body": "",
        }

    # パスに基づいて処理
    if "/multimodal" in path:
        return handle_multimodal(event)
    elif "/voice" in path:
        return handle_voice(event)
    else:
        return response(404, {"error": "Not found"})


def handle_multimodal(event):
    """
    Multimodal エージェント処理
    
    Bedrock Nova Lite/Pro を使用して画像解析を実行
    """
    http_method = event.get("httpMethod", "GET")

    if http_method == "POST":
        try:
            body = json.loads(event.get("body", "{}"))
            message = body.get("message", "この画像を説明してください")
            images = body.get("images", [])
            mode = body.get("mode", "understand")  # understand, generate, video_understand, video_generate

            logger.info(f"Multimodal request: mode={mode}, images={len(images)}, message={message[:50]}")

            # モード別処理
            if mode == "understand" or mode == "画像理解":
                return handle_image_understanding(message, images)
            elif mode == "generate" or mode == "画像生成":
                return handle_image_generation(message)
            elif mode == "video_understand" or mode == "動画理解":
                return handle_video_understanding(message, images)
            elif mode == "video_generate" or mode == "動画生成":
                return handle_video_generation(message)
            else:
                # デフォルトは画像理解
                return handle_image_understanding(message, images)

        except Exception as e:
            logger.exception(f"Multimodal error: {e}")
            return response(500, {"error": str(e), "type": type(e).__name__})

    return response(405, {"error": "Method not allowed"})


def handle_image_understanding(message: str, images: list) -> dict:
    """
    画像理解 (Nova Lite/Pro Vision)
    
    画像をアップロードしてAIに内容を説明させる
    """
    if not images:
        return response(400, {"error": "No images provided for understanding"})

    try:
        # メッセージコンテンツを構築
        content = []
        
        # 画像を追加
        for img in images:
            if isinstance(img, dict):
                # Base64形式
                image_data = img.get("data", img.get("base64", ""))
                media_type = img.get("mediaType", img.get("type", "image/jpeg"))
            elif isinstance(img, str):
                # 直接Base64文字列
                image_data = img
                media_type = "image/jpeg"
            else:
                continue
            
            # Base64プレフィックスを除去
            if "," in image_data:
                image_data = image_data.split(",")[1]
            
            content.append({
                "image": {
                    "format": media_type.split("/")[-1] if "/" in media_type else "jpeg",
                    "source": {
                        "bytes": base64.b64decode(image_data)
                    }
                }
            })
        
        # テキストプロンプトを追加
        content.append({
            "text": message or "この画像の内容を詳しく説明してください。"
        })

        # Bedrock Converse API 呼び出し
        response_data = bedrock_runtime.converse(
            modelId=NOVA_LITE_MODEL,
            messages=[
                {
                    "role": "user",
                    "content": content
                }
            ],
            inferenceConfig={
                "maxTokens": 2048,
                "temperature": 0.7,
            },
            system=[
                {
                    "text": "あなたは画像分析の専門家です。画像の内容を詳細に分析し、日本語で分かりやすく説明してください。"
                }
            ]
        )

        # レスポンス抽出
        output_message = response_data.get("output", {}).get("message", {})
        output_content = output_message.get("content", [])
        
        result_text = ""
        for item in output_content:
            if "text" in item:
                result_text += item["text"]

        usage = response_data.get("usage", {})

        logger.info(f"Image understanding completed: {len(result_text)} chars")

        return response(200, {
            "status": "success",
            "mode": "image_understanding",
            "model": NOVA_LITE_MODEL,
            "response": result_text,
            "usage": {
                "inputTokens": usage.get("inputTokens", 0),
                "outputTokens": usage.get("outputTokens", 0),
            },
            "metadata": {
                "images_processed": len(images),
                "prompt": message,
            }
        })

    except Exception as e:
        logger.exception(f"Image understanding error: {e}")
        return response(500, {
            "status": "error",
            "error": str(e),
            "type": type(e).__name__,
            "message": "画像解析中にエラーが発生しました"
        })


def handle_image_generation(prompt: str) -> dict:
    """
    画像生成 (Nova Canvas)
    
    テキストプロンプトから画像を生成
    """
    try:
        # Nova Canvas モデル
        NOVA_CANVAS_MODEL = os.environ.get("NOVA_CANVAS_MODEL", "amazon.nova-canvas-v1:0")
        
        # 画像生成リクエスト
        request_body = {
            "taskType": "TEXT_IMAGE",
            "textToImageParams": {
                "text": prompt,
            },
            "imageGenerationConfig": {
                "numberOfImages": 1,
                "width": 1024,
                "height": 1024,
                "cfgScale": 8.0,
                "seed": 0,
            }
        }

        response_data = bedrock_runtime.invoke_model(
            modelId=NOVA_CANVAS_MODEL,
            body=json.dumps(request_body),
            contentType="application/json",
            accept="application/json"
        )

        result = json.loads(response_data["body"].read())
        images = result.get("images", [])

        if images:
            return response(200, {
                "status": "success",
                "mode": "image_generation",
                "model": NOVA_CANVAS_MODEL,
                "images": images,  # Base64エンコードされた画像
                "metadata": {
                    "prompt": prompt,
                    "count": len(images),
                }
            })
        else:
            return response(200, {
                "status": "no_images",
                "message": "画像を生成できませんでした",
                "prompt": prompt,
            })

    except Exception as e:
        logger.exception(f"Image generation error: {e}")
        return response(500, {
            "status": "error",
            "error": str(e),
            "message": "画像生成中にエラーが発生しました"
        })


def handle_video_understanding(message: str, videos: list) -> dict:
    """
    動画理解 (Nova Pro Vision)
    
    動画をアップロードしてAIに内容を説明させる
    """
    if not videos:
        return response(400, {"error": "No videos provided for understanding"})

    try:
        # 動画の最初のフレームまたはサムネイルを使用
        # 注: Nova Proは動画の直接処理に制限があるため、フレーム抽出が推奨
        
        content = []
        
        for video in videos:
            if isinstance(video, dict):
                video_data = video.get("data", video.get("base64", ""))
                # 動画の場合、最初のフレームを抽出する処理が必要
                # ここではサムネイルとして扱う
            elif isinstance(video, str):
                video_data = video
            else:
                continue
            
            if "," in video_data:
                video_data = video_data.split(",")[1]
            
            content.append({
                "image": {
                    "format": "jpeg",
                    "source": {
                        "bytes": base64.b64decode(video_data)
                    }
                }
            })
        
        content.append({
            "text": message or "この動画/フレームの内容を説明してください。"
        })

        response_data = bedrock_runtime.converse(
            modelId=NOVA_PRO_MODEL,
            messages=[
                {
                    "role": "user",
                    "content": content
                }
            ],
            inferenceConfig={
                "maxTokens": 2048,
                "temperature": 0.7,
            },
            system=[
                {
                    "text": "あなたは動画分析の専門家です。動画/フレームの内容を詳細に分析し、日本語で分かりやすく説明してください。"
                }
            ]
        )

        output_message = response_data.get("output", {}).get("message", {})
        output_content = output_message.get("content", [])
        
        result_text = ""
        for item in output_content:
            if "text" in item:
                result_text += item["text"]

        return response(200, {
            "status": "success",
            "mode": "video_understanding",
            "model": NOVA_PRO_MODEL,
            "response": result_text,
            "metadata": {
                "frames_processed": len(videos),
                "prompt": message,
            }
        })

    except Exception as e:
        logger.exception(f"Video understanding error: {e}")
        return response(500, {
            "status": "error",
            "error": str(e),
            "message": "動画解析中にエラーが発生しました"
        })


def handle_video_generation(prompt: str) -> dict:
    """
    動画生成 (Nova Reel)
    
    テキストプロンプトから動画を生成
    """
    try:
        # Nova Reel モデル
        NOVA_REEL_MODEL = os.environ.get("NOVA_REEL_MODEL", "amazon.nova-reel-v1:0")
        
        # 動画生成は非同期処理が推奨
        # ここでは開始のみを行い、結果はS3に出力
        
        return response(200, {
            "status": "pending",
            "mode": "video_generation",
            "model": NOVA_REEL_MODEL,
            "message": "動画生成は非同期処理です。生成完了後、S3に出力されます。",
            "info": {
                "prompt": prompt,
                "note": "動画生成は時間がかかるため、バックグラウンドで処理されます。"
            }
        })

    except Exception as e:
        logger.exception(f"Video generation error: {e}")
        return response(500, {
            "status": "error",
            "error": str(e),
            "message": "動画生成の開始中にエラーが発生しました"
        })


def handle_voice(event):
    """
    Voice エージェント処理
    
    Nova Sonic を使用した音声処理
    """
    http_method = event.get("httpMethod", "GET")
    path = event.get("path", "")

    if http_method == "POST":
        try:
            body = json.loads(event.get("body", "{}"))

            if "/process" in path:
                audio = body.get("audio", "")
                
                # 注: Nova Sonic は双方向ストリーミングが必要
                # Lambda単体では完全な実装が困難
                # AgentCore Runtime での実装を推奨
                
                return response(200, {
                    "status": "info",
                    "message": "Voice processing with Nova Sonic requires WebSocket streaming.",
                    "recommendation": "Use AgentCore Runtime with WebSocket support for real-time voice dialogue.",
                    "alternative": {
                        "transcription": "Use Amazon Transcribe for speech-to-text",
                        "synthesis": "Use Amazon Polly for text-to-speech"
                    }
                })
                
            elif "/text" in path:
                text = body.get("text", "")
                
                # テキスト処理は Bedrock で可能
                response_data = bedrock_runtime.converse(
                    modelId=NOVA_LITE_MODEL,
                    messages=[
                        {
                            "role": "user",
                            "content": [{"text": text}]
                        }
                    ],
                    inferenceConfig={
                        "maxTokens": 1024,
                        "temperature": 0.7,
                    }
                )
                
                output_message = response_data.get("output", {}).get("message", {})
                output_content = output_message.get("content", [])
                
                result_text = ""
                for item in output_content:
                    if "text" in item:
                        result_text += item["text"]
                
                return response(200, {
                    "status": "success",
                    "response": result_text,
                    "input": text,
                })
            else:
                return response(200, {
                    "status": "ok",
                    "endpoints": {
                        "/voice/process": "Audio processing (requires WebSocket)",
                        "/voice/text": "Text processing"
                    }
                })

        except Exception as e:
            logger.exception(f"Voice error: {e}")
            return response(500, {"error": str(e)})

    return response(405, {"error": "Method not allowed"})


def response(status_code: int, body: dict) -> dict:
    """レスポンス生成"""
    return {
        "statusCode": status_code,
        "headers": {
            **CORS_HEADERS,
            "Content-Type": "application/json",
        },
        "body": json.dumps(body, ensure_ascii=False),
    }

File: amplify/functions/agent-resolver/handler.ts
/**
 * Agent Resolver Lambda Handler
 *
 * AgentCore + StrandsAgents + Bedrock Nova シリーズを使用した
 * Multimodal/Voice エージェント処理
 *
 * 設計原則:
 * - AgentCore + StrandsAgents + BedrockAPI 構成
 * - AgentCore_Observability / CloudTrail 追跡可能
 * - boto3 / cli / script / sh 直接処理禁止
 * - 12 Agent Factor 準拠
 */

import type { AppSyncResolverHandler } from 'aws-lambda';
import {
  BedrockRuntimeClient,
  InvokeModelCommand,
  ConverseCommand,
} from '@aws-sdk/client-bedrock-runtime';

// Types aligned with Domain Layer
interface MultimodalResponse {
  message?: string;
  images?: Array<{ base64: string; seed?: number }>;
  videos?: Array<{ url: string }>;
  metadata?: Record<string, unknown>;
}

interface VoiceResponse {
  transcript?: string;
  userText?: string;
  assistantText?: string;
  audio?: string; // Base64 encoded
  metadata?: Record<string, unknown>;
}

interface InvokeMultimodalArgs {
  sessionId: string;
  prompt: string;
  image?: string; // Base64 encoded input image
}

interface SendVoiceTextArgs {
  sessionId: string;
  text: string;
}

interface InvokeVoiceAgentArgs {
  sessionId: string;
  text?: string;
  audio?: string; // Base64 encoded audio
  mode?: 'TTS' | 'STT' | 'DIALOGUE';
}

type ResolverArgs = InvokeMultimodalArgs | SendVoiceTextArgs | InvokeVoiceAgentArgs;

// Environment
const REGION = process.env.AWS_REGION || 'ap-northeast-1';
const AGENTCORE_MEMORY_ID = process.env.AGENTCORE_MEMORY_ID || '';
const OUTPUT_BUCKET = process.env.OUTPUT_BUCKET || '';
const LOG_LEVEL = process.env.LOG_LEVEL || 'INFO';

// Model IDs
// Note: Nova models require inference profiles for on-demand invocation
// Using Claude models for text generation (no inference profile required)
const TEXT_MODEL_ID = 'anthropic.claude-3-haiku-20240307-v1:0'; // 会話用
const NOVA_CANVAS_MODEL_ID = 'amazon.nova-canvas-v1:0'; // 画像生成用 (inference profile 必要)
const NOVA_REEL_MODEL_ID = 'amazon.nova-reel-v1:0'; // 動画生成用 (inference profile 必要)
const NOVA_SONIC_MODEL_ID = 'amazon.nova-sonic-v1:0'; // 音声用 (inference profile 必要)

// Bedrock Client
const bedrockClient = new BedrockRuntimeClient({ region: REGION });

// Observability: Structured logging
const log = (level: string, message: string, data?: Record<string, unknown>) => {
  const logLevels = { DEBUG: 0, INFO: 1, WARN: 2, ERROR: 3 };
  const currentLevel = logLevels[LOG_LEVEL as keyof typeof logLevels] ?? 1;
  const messageLevel = logLevels[level as keyof typeof logLevels] ?? 1;

  if (messageLevel >= currentLevel) {
    console.log(
      JSON.stringify({
        timestamp: new Date().toISOString(),
        level,
        message,
        ...data,
        service: 'agent-resolver',
      })
    );
  }
};

export const handler: AppSyncResolverHandler<
  ResolverArgs,
  MultimodalResponse | VoiceResponse
> = async (event) => {
  // Amplify Gen2 handler structure: fieldName is directly on event
  const fieldName = (event as any).fieldName || event.info?.fieldName;
  const args = (event as any).arguments || event.arguments;

  log('INFO', 'Agent Resolver invoked', {
    fieldName,
    eventKeys: Object.keys(event),
  });

  try {
    switch (fieldName) {
      case 'invokeMultimodal':
        return await invokeMultimodal(args as InvokeMultimodalArgs);
      case 'sendVoiceText':
        return await sendVoiceText(args as SendVoiceTextArgs);
      case 'invokeVoiceAgent':
        return await invokeVoiceAgent(args as InvokeVoiceAgentArgs);
      default:
        throw new Error(`Unknown field: ${fieldName}`);
    }
  } catch (error) {
    log('ERROR', 'Agent Resolver failed', {
      error: error instanceof Error ? error.message : String(error),
      fieldName,
    });
    throw error;
  }
};

/**
 * Multimodal Agent 呼び出し
 * Nova Pro/Lite + Canvas + Reel を組み合わせ
 */
async function invokeMultimodal(args: InvokeMultimodalArgs): Promise<MultimodalResponse> {
  const { sessionId, prompt, image } = args;

  // プロンプトから制御シグナルを除去（表示用）
  const cleanPrompt = prompt
    .replace('[IMAGE_GENERATION]', '')
    .replace('[VIDEO_GENERATION]', '')
    .trim();

  const isImageGeneration = shouldGenerateImage(prompt);
  const isVideoGeneration = shouldGenerateVideo(prompt);

  log('INFO', 'Invoking multimodal agent', { 
    sessionId, 
    hasImage: !!image,
    isImageGeneration,
    isVideoGeneration,
    promptLength: prompt.length
  });

  const response: MultimodalResponse = {
    metadata: {
      sessionId,
      timestamp: new Date().toISOString(),
    },
  };

  try {
    // 画像生成モードの場合
    if (isImageGeneration) {
      log('INFO', 'Processing image generation request', { cleanPrompt });
      
      const imageResult = await invokeNovaCanvas(cleanPrompt);
      if (imageResult) {
        response.images = [imageResult];
        response.message = `「${cleanPrompt}」の画像を生成しました。`;
      } else {
        // Nova Canvas が失敗した場合、テキストで説明を生成
        response.message = `画像生成に失敗しました。Nova Canvas モデルへのアクセスを確認してください。プロンプト: "${cleanPrompt}"`;
      }
      return response;
    }

    // 動画生成モードの場合
    if (isVideoGeneration) {
      log('INFO', 'Processing video generation request', { cleanPrompt });
      
      // Nova Reel は非同期処理のため、ステータス URL を返す
      response.videos = [{ url: 'pending://nova-reel-job-id' }];
      response.message = `動画生成ジョブを開始しました。プロンプト: "${cleanPrompt}"`;
      return response;
    }

    // 通常の会話/画像解析モード
    const converseResponse = await invokeNovaConverse(cleanPrompt || prompt, image);
    response.message = converseResponse;

    log('INFO', 'Multimodal response generated', {
      hasMessage: !!response.message,
      imageCount: response.images?.length || 0,
      videoCount: response.videos?.length || 0,
    });

    return response;
  } catch (error) {
    log('ERROR', 'Multimodal processing failed', {
      error: error instanceof Error ? error.message : String(error),
    });

    // Graceful degradation
    return {
      message: 'Processing error. Please try again.',
      metadata: {
        error: true,
        sessionId,
      },
    };
  }
}

/**
 * Voice Agent (Text-to-Speech/Speech-to-Text)
 * Nova Sonic を使用
 */
async function sendVoiceText(args: SendVoiceTextArgs): Promise<VoiceResponse> {
  const { sessionId, text } = args;

  log('INFO', 'Processing voice text', { sessionId, textLength: text.length });

  try {
    // Nova Pro で応答テキスト生成
    const assistantText = await invokeNovaConverse(text);

    // TODO: Nova Sonic で音声合成（WebSocket 経由）
    // 現在はテキストレスポンスのみ

    const response: VoiceResponse = {
      userText: text,
      assistantText,
      metadata: {
        sessionId,
        model: NOVA_SONIC_MODEL_ID,
        timestamp: new Date().toISOString(),
      },
    };

    log('INFO', 'Voice response generated', { sessionId });
    return response;
  } catch (error) {
    log('ERROR', 'Voice processing failed', {
      error: error instanceof Error ? error.message : String(error),
    });

    return {
      userText: text,
      assistantText: 'I apologize, but I encountered an error processing your request.',
      metadata: {
        error: true,
        sessionId,
      },
    };
  }
}

/**
 * Voice Agent with audio input support
 * Supports TTS, STT, and full dialogue modes
 */
async function invokeVoiceAgent(args: InvokeVoiceAgentArgs): Promise<VoiceResponse> {
  const { sessionId, text, audio, mode = 'DIALOGUE' } = args;

  log('INFO', 'Invoking voice agent', { 
    sessionId, 
    mode, 
    hasText: !!text, 
    hasAudio: !!audio 
  });

  try {
    let transcription: string | undefined;
    let responseText: string | undefined;

    // Step 1: Speech-to-Text if audio provided
    if (audio && (mode === 'STT' || mode === 'DIALOGUE')) {
      log('DEBUG', 'Processing audio input for STT');
      // TODO: Use Nova Sonic or Amazon Transcribe for STT
      // For now, return placeholder
      transcription = '(音声認識は開発中です)';
    }

    // Step 2: Generate response text
    const inputText = text || transcription || '';
    if (inputText && (mode === 'TTS' || mode === 'DIALOGUE')) {
      responseText = await invokeNovaConverse(inputText);
    }

    // Step 3: Text-to-Speech if needed
    let responseAudio: string | undefined;
    if (responseText && (mode === 'TTS' || mode === 'DIALOGUE')) {
      // TODO: Use Nova Sonic or Amazon Polly for TTS
      // For now, no audio output
      log('DEBUG', 'TTS generation skipped (not yet implemented)');
    }

    const response: VoiceResponse = {
      transcript: transcription,
      userText: text,
      assistantText: responseText || (mode === 'STT' ? '文字起こし完了' : ''),
      audio: responseAudio,
      metadata: {
        sessionId,
        mode,
        timestamp: new Date().toISOString(),
      },
    };

    log('INFO', 'Voice agent response generated', { 
      sessionId, 
      hasTranscription: !!transcription,
      hasResponseText: !!responseText 
    });

    return response;
  } catch (error) {
    log('ERROR', 'Voice agent processing failed', {
      error: error instanceof Error ? error.message : String(error),
    });

    return {
      assistantText: 'エラーが発生しました。もう一度お試しください。',
      metadata: {
        error: true,
        sessionId,
        mode,
      },
    };
  }
}

/**
 * Nova Pro/Lite Converse API 呼び出し（画像分析対応）
 */
async function invokeNovaConverse(prompt: string, image?: string): Promise<string> {
  // Build message content
  const content: Array<{ text: string } | { image: { format: 'png' | 'jpeg' | 'gif' | 'webp'; source: { bytes: Uint8Array } } }> = [];
  
  // Add text prompt
  content.push({ text: prompt || 'この画像を詳しく説明してください。' });
  
  // Add image if provided (convert Base64 to Uint8Array)
  if (image) {
    try {
      // Decode Base64 to binary
      const binaryString = atob(image);
      const bytes = new Uint8Array(binaryString.length);
      for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
      }
      
      // Detect image format from first bytes
      const format = detectImageFormat(bytes);
      
      content.push({
        image: {
          format,
          source: { bytes },
        },
      });
      
      log('DEBUG', 'Image attached to request', { 
        format, 
        sizeBytes: bytes.length 
      });
    } catch (error) {
      log('ERROR', 'Failed to decode image', { 
        error: error instanceof Error ? error.message : String(error) 
      });
    }
  }

  const messages = [
    {
      role: 'user' as const,
      content,
    },
  ];

  log('DEBUG', 'Sending Converse request', { 
    modelId: TEXT_MODEL_ID,
    hasImage: !!image,
    contentParts: content.length
  });

  const command = new ConverseCommand({
    modelId: TEXT_MODEL_ID,
    messages: messages as any,
    inferenceConfig: {
      maxTokens: 2048,
      temperature: 0.7,
    },
  });

  const response = await bedrockClient.send(command);
  const outputMessage = response.output?.message;

  if (outputMessage?.content?.[0] && 'text' in outputMessage.content[0]) {
    return outputMessage.content[0].text || '';
  }

  return '';
}

/**
 * Detect image format from binary data
 */
function detectImageFormat(bytes: Uint8Array): 'png' | 'jpeg' | 'gif' | 'webp' {
  // PNG: 89 50 4E 47
  if (bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4E && bytes[3] === 0x47) {
    return 'png';
  }
  // JPEG: FF D8 FF
  if (bytes[0] === 0xFF && bytes[1] === 0xD8 && bytes[2] === 0xFF) {
    return 'jpeg';
  }
  // GIF: 47 49 46
  if (bytes[0] === 0x47 && bytes[1] === 0x49 && bytes[2] === 0x46) {
    return 'gif';
  }
  // WebP: 52 49 46 46 ... 57 45 42 50
  if (bytes[0] === 0x52 && bytes[1] === 0x49 && bytes[2] === 0x46 && bytes[3] === 0x46 &&
      bytes[8] === 0x57 && bytes[9] === 0x45 && bytes[10] === 0x42 && bytes[11] === 0x50) {
    return 'webp';
  }
  // Default to jpeg (most common)
  return 'jpeg';
}

/**
 * Nova Canvas で画像生成
 */
async function invokeNovaCanvas(
  prompt: string
): Promise<{ base64: string; seed: number } | null> {
  log('INFO', 'Starting Nova Canvas image generation', {
    modelId: NOVA_CANVAS_MODEL_ID,
    promptLength: prompt.length,
  });

  try {
    const requestBody = {
      taskType: 'TEXT_IMAGE',
      textToImageParams: {
        text: prompt,
      },
      imageGenerationConfig: {
        numberOfImages: 1,
        width: 1024,
        height: 1024,
        cfgScale: 8.0,
      },
    };

    log('DEBUG', 'Nova Canvas request body', { requestBody });

    const command = new InvokeModelCommand({
      modelId: NOVA_CANVAS_MODEL_ID,
      contentType: 'application/json',
      accept: 'application/json',
      body: JSON.stringify(requestBody),
    });

    const response = await bedrockClient.send(command);
    const responseBody = JSON.parse(new TextDecoder().decode(response.body));

    log('DEBUG', 'Nova Canvas response', {
      hasImages: !!responseBody.images,
      imageCount: responseBody.images?.length || 0,
    });

    if (responseBody.images?.[0]) {
      log('INFO', 'Nova Canvas image generation successful');
      return {
        base64: responseBody.images[0],
        seed: responseBody.seeds?.[0] || 0,
      };
    }

    log('WARN', 'Nova Canvas returned no images', { responseBody });
    return null;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    const errorName = error instanceof Error ? error.name : 'Unknown';
    
    log('ERROR', 'Nova Canvas generation failed', {
      errorName,
      errorMessage,
      modelId: NOVA_CANVAS_MODEL_ID,
      stack: error instanceof Error ? error.stack : undefined,
    });
    return null;
  }
}

/**
 * 画像生成が必要かどうかを判定
 */
function shouldGenerateImage(prompt: string): boolean {
  const imageKeywords = [
    '[IMAGE_GENERATION]', // フロントエンドからのシグナル
    '画像を生成',
    '画像を作成',
    'generate image',
    'create image',
    'draw',
    '描いて',
    'イラスト',
  ];
  const lowerPrompt = prompt.toLowerCase();
  return imageKeywords.some((keyword) => lowerPrompt.includes(keyword.toLowerCase()));
}

/**
 * 動画生成が必要かどうかを判定
 */
function shouldGenerateVideo(prompt: string): boolean {
  const videoKeywords = [
    '[VIDEO_GENERATION]', // フロントエンドからのシグナル
    '動画を生成',
    '動画を作成',
    'generate video',
    'create video',
    'アニメーション',
  ];
  const lowerPrompt = prompt.toLowerCase();
  return videoKeywords.some((keyword) => lowerPrompt.includes(keyword.toLowerCase()));
}
